appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroBase.java

```diff
@@ -1052,36 +1118,129 @@ private void updatePagerIndicatorState() {
         }
     }
 
+    /**
+     * Empty method that points to {@link #askForPermissions(String[], int, boolean)}
+     *
+     * @param permissions
+     * @param slidesNumber
+     */
     public void askForPermissions(String[] permissions, int slidesNumber) {
+        askForPermissions(permissions, slidesNumber, false);
+    }
+
+    /**
+     * Called by the user to associate permissions with slides.
+     *
+     * @param permissions  - Array of permissions
+     * @param slidesNumber - The slide at which permission is to be asked.
+     * @param required     - Whether the user can change this slide without granting the permissions.
+     */
+    public void askForPermissions(String[] permissions, int slidesNumber, boolean required) {
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
             if (slidesNumber == 0) {
                 Toast.makeText(getBaseContext(), "Invalid Slide Number", Toast.LENGTH_SHORT).show();
             } else {
-                PermissionWrapper permission = new PermissionWrapper(permissions, slidesNumber);
-                permissionsArray.add(permission);
-                setSwipeLock(true);
+                PermissionWrapper permission = new PermissionWrapper(permissions, slidesNumber, required);
+                if (!permissionsArray.contains(permission)) {
+                    permissionsArray.add(permission);
+                    permissionsPositions.add(slidesNumber);// Add the number of the slide that contains permission to an arraylist
+                }
             }
         }
     }
 
     public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[],
                                            @NonNull int[] grantResults) {
         super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+        setSwipeLock(false);
+        pager.setPermDialogSwipeLastCalled(System.currentTimeMillis());
```

Comment:
No, it isn't. Look, if we have permissions on 2 slides, I request permissions on one, press allow/deny l, then quickly swipe again such that the next slide is only halfway there and the permission dialogue opens, so this call prevent the swipe from doing anything before 200 ms.

Reply:
> Look, if we have permissions on 2 slides, I request permissions on one, press allow/deny l, then quickly swipe again such that the next slide is only halfway there and the permission dialogue opens, so this call prevent the swipe from doing anything before 200 ms.

I don't fully understand sorry :|

My point is `setPermDialogSwipeLastCalled`, will save the last time that `PermDialogSwipe` was called. But here you're inside the `onRequestPermissionsResult`. Why?

Reply:
Because, after we click on allow/deny, intermissions result gets called and the slide gets changed if required. Imagine this: 
Permission on 2nd and 3rd slide. We click allow on slide 2 and then quickly swipe forward, so the permission dialog opens while slide 3 is halfway there. This call ensures that the permission dialog can't open before 200ms of clicking allow/deny and gives slide3 enough time to settle down

Reply:
@cortinico Setting permDialogswipelastcalled is just a way to prevent user input while the viewpager is changing slides.